src/main/scala/workflow/graph/AutoCacheRule.scala

```diff
@@ -0,0 +1,789 @@
+    runs: Map[NodeId, Int],
+    spaceLeft: Long
+  ): NodeId = {
+    //Get the uncached node which fits that maximizes savings in runtime.
+    graph.nodes.filter(node =>
+      !cached(node) &&
+        profiles.getOrElse(node, Profile(0, 0, 0)).rddMem < spaceLeft &&
+        runs(node) > 1)
+      .minBy(i => estimateCachedRunTime(graph, linearization, childrenByNode, cached + i, profiles))
+  }
+
+  def greedyCache(
+    graph: Graph,
+    profiles: Map[NodeId, Profile],
+    maxMem: Option[Long]
+  ): Graph = {
+    val cached = initCacheSet(graph)
+    val nodeWeights = getNodeWeights(graph)
+    val childrenByNode = getChildrenForAllNodes(graph)
+    val linearization = AnalysisUtils.linearize(graph)
+    var runs = getRuns(linearization, childrenByNode, cached, nodeWeights)
+    val descendantsOfSources = getDescendantsOfSources(graph)
+
+    var toCache = Set[NodeId]()
+    val memBudget = maxMem.getOrElse(graph.operators.values.collectFirst {
+      // Collect the sparkContext from any DatasetOperators, and find what is already cached
+      case DatasetOperator(rdd) =>
+        val sc = rdd.sparkContext
+        // Calculate memory budget, ignoring driver unless in local mode
+        val workers = if (sc.getExecutorStorageStatus.length == 1) {
+          sc.getExecutorStorageStatus
+        } else {
+          sc.getExecutorStorageStatus.filter(w => !w.blockManagerId.isDriver)
+        }
+
+        val totalMem = workers.map(_.memRemaining.toDouble).sum
+        totalMem * 0.75
+    }.getOrElse(0.0).toLong)
+
+    var usedMem = cachedMem(cached, profiles)
+    while (usedMem < memBudget && stillRoom(cached ++ toCache, runs, profiles, memBudget - usedMem)) {
+      toCache = toCache + selectNext(
+        graph,
+        linearization,
+        profiles,
+        childrenByNode,
+        cached ++ toCache,
+        runs,
+        memBudget - usedMem)
+      runs = getRuns(linearization, childrenByNode, cached ++ toCache, nodeWeights)
+      usedMem = cachedMem(cached ++ toCache, profiles)
+    }
+
+    addCachesToPipeline(graph, toCache -- descendantsOfSources)
+  }
+
+  override def apply(plan: Graph, prefixes: Map[NodeId, Prefix]): (Graph, Map[NodeId, Prefix]) = {
+    cachingMode match {
+      case AggressiveCache => (aggressiveCache(plan), prefixes)
+      case GreedyCache(maxMem, profileScales, numProfileTrials) => {
+        logInfo("Starting pipeline profile")
+        val cached = initCacheSet(plan)
+        val nodeWeights = getNodeWeights(plan)
+        val childrenByNode = getChildrenForAllNodes(plan)
+        val linearization = AnalysisUtils.linearize(plan)
+        val runs = getRuns(linearization, childrenByNode, cached, nodeWeights)
+        val descendantsOfSources = getDescendantsOfSources(plan)
+
+        // We have the possibility of caching all uncached nodes accessed more than once,
+        // That don't depend on the sources.
+        // TODO: MUST profile all Cacher nodes!!! (to get cachedMem to be correct)
```

Comment:
No, not yet. I wasn't sure if you wanted to profile all nodes marked cached (to keep track of driver mem), or just the cached rdds to get an accurate cluster mem.